File: DSA/Graphs/DepthFirstSearch/code/dfs.py

'''
The graph is represented as adjacency lists stored in a dictionary. It is a directed graph.
'''

graph = {
    0: [1, 2],
    1: [0, 3, 4],
    2: [0, 4],
    3: [1],
    4: [1, 2]
}



def dfs(graph, source):
    
    visited = set()
    stack = [source] #Add the source node to the stack

    while stack:
        v = stack.pop() 
        if v not in visited: #If v has not been visited, 
            visited.add(v)
            for u in graph[v]:
                if u not in visited:
                    stack.append(u)
    
    return visited #Returns the set of all visited nodes in the graph


def dfs_recursive(graph, source, visited):

    if source not in visited:
        visited.add(source)
        for u in graph[source]:
            if u not in visited:
                dfs_recursive(graph, u, visited)
    return visited


print(f"DFS iterative: {dfs(graph, 0)}")
print(f"DFS recursive: {dfs_recursive(graph, 0, set())}")

File: Data Structures and Algorithms/Graphs/DepthFirstSearch/code/DFS.java
package DepthFirstSearch.code;

import java.util.ArrayDeque;
import java.util.Deque;

/* The graph is a directed graph represented as an adjacency matrix */

/**
 * dfs
 */
public class DFS {

    public static int [][] graph = {
   
        {0, 1, 1, 0, 1},  /// Node 0 has edges to nodes 1, 2, and 4
        {1, 0, 1, 0, 0},  ///Node 1 has edges to nodes 0 and 2
        {0, 1, 0, 1, 1},  /// Node 2 has edges to nodes 1, 3, and 4
        {1, 0, 0, 0, 1},  /// Node 3 has edges to nodes 0 and 4
        {0, 1, 1, 1, 0} /// Node 4 has edges to nodes 1, 2, and 3
    };

    public static boolean[] dfs(int[][] graph, int source){
        
        boolean[] visited = new boolean[graph.length];
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(source);

        while (stack.peek() != null){
            int v = stack.pop();
            if (!visited[v]){
                visited[v] = true;
                for (int u = 0; u < graph[v].length; u++) {
                    if (graph[v][u] == 1 && !visited[u]) {
                        stack.push(u);
                    }
                }
            }
        }
        return visited;
    }

    public static boolean[] dfsRecursive(int[][] graph, int source, boolean[] visited){
        if (!visited[source]){
            visited[source] = true;
            for (int u = 0; u < graph[source].length; u++){
                if (graph[source][u] == 1 && !visited[u]){
                    dfsRecursive(graph, u, visited);
                }
            }
        }
        return visited;
    }

    public static void main(String[] args) {
        boolean[] dfsRes = dfs(graph, 0);

        boolean[] visited = new boolean[graph.length];
        boolean[] dfsRecRes = dfsRecursive(graph, 0, visited);

        System.out.print("DFS iterative: ");
        for (boolean value: dfsRes){
            System.out.printf("%s, ",value);
        }
        System.out.println();

        System.out.print("DFS recursive: ");
        for (boolean value: dfsRecRes){
            System.out.printf("%s, ",value);
        }

       
    }
    
}


